core/src/main/scala/kafka/server/FetchSession.scala

```diff
@@ -71,49 +71,46 @@ object FetchSession {
   * Note that fetcherLogStartOffset is the LSO of the follower performing the fetch, whereas
   * localLogStartOffset is the log start offset of the partition on this broker.
   */
-class CachedPartition(val topic: String,
-                      val partition: Int,
+class CachedPartition(val topicPartition: TopicPartition,
                       var maxBytes: Int,
                       var fetchOffset: Long,
                       var highWatermark: Long,
+                      var leaderEpoch: Optional[Integer],
```

Comment:
It seems like rather than use None for older fetch requests, we could just initialize the leaderEpoch to the current leader epoch.  That way, we could at least detect epoch changes.  Although we'd have to figure out how to handle the error-- probably just by dropping the session.

Reply:
Can you elaborate a little bit? I think the validation only has value if it comes from the client making the request.

Reply:
Discussed offline. It's an interesting thought. The problem is we don't have an error that would force a follower to go through the truncation phase. We could force a retry and we could drop the session, but the follower would still be able to continue.

Reply:
Can we file a separate JIRA for this idea?

Reply:
Thinking about this more, I'm not sure it's worth it.  Without true leader epoch awareness in the client, we don't have a way to force the client to refresh metadata and use a different leader, which is what we would need here.  So maybe my initial idea was a bit half-baked-- sorry.